### PR TITLE
PP-3579 - AB testing - Firebreak

### PR DESCRIPTION
**Pull request by tlwr**
# A/B Testing

- Added middleware to split traffic between two routes
- Added analytics so we can measure which one they used
- Amended tests and set up env vars and query params so we can disable if we want to

### CODE REVIEW

**Comment by brid** [comment]
I've spoken with Jon about this and I think we need to have a level of confidence that any cookies this generates, don't risk affecting payment journeys, especially given we have multiple charges being attempted per session.

Knowing the lower/upper bounds on what affect this has on cookie sizes would get us most of the way there, preferably through monitoring this via Chrome developer tools or something similar.

**Comment by tlwr** [comment]
This will add a cookie with a small key and a small value

It might cause some 502s but that is because of the chargeId cookie dropping not the AB testing cookie dropping

We should fix the chargeID cookie dropping, and merge this concurrently

File: app/router.js [hunk 1]
```diff
@@ -52,7 +54,15 @@ exports.bind = function (app) {
   app.post(card.auth3dsHandler.path, middlewareStack, charge.auth3dsHandler)
   app.get(card.captureWaiting.path, middlewareStack, charge.captureWaiting)
   app.post(card.create.path, middlewareStack, charge.create)
-  app.get(card.confirm.path, middlewareStack, charge.confirm)
+  app.get(
+    card.confirm.path,
+    middlewareStack,
+    abTest.switch({
+      threshold: process.env.AB_TEST_THRESHOLD || 100,
```
Comment (maxcbc) [inline app/router.js]: As a general rule I would put this up top as a constant. 
```
const AB_TEST_THRESHOLD = process.env.AB_TEST_THRESHOLD || 100,
```

File: app/utils/ab_test.js [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+exports.switch = function (opts) {
+  let threshold = Math.floor(opts.threshold || 90)
+  let defaultVariant = opts.defaultVariant
+  let testingVariant = opts.testingVariant
+
+  return function (req, res) {
+    let sessionValue = exports._getOrSetSession(req)
+
+    let showDefault = true
```
Comment (maxcbc) [inline app/utils/ab_test.js]: why is this explicitly true? why this variable?

File: app/utils/ab_test.js [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+  if (exports._getSession(req) === null) {
+    logger.info('Could not find user ab testing session, creating one')
+    exports._setSession(req)
+  }
+  return exports._getSession(req)
+}
+
+exports.switch = function (opts) {
+  let threshold = Math.floor(opts.threshold || 90)
+  let defaultVariant = opts.defaultVariant
```
Comment (maxcbc) [inline app/utils/ab_test.js]: most of these can be `const` as they are never changed

File: app/utils/ab_test.js [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+exports._setSession = function (req) {
+  return cookie.setSessionVariable(
+    req,
+    'abTestId',
+    exports.uniformlyGeneratedRandomNumber()
+  )
+}
+
+exports._getOrSetSession = function (req) {
```
Comment (maxcbc) [inline app/utils/ab_test.js]: slight style point, you can use fat-arrow `()=>{}` functions for all these.

File: app/utils/analytics.js [hunk 5]
```diff
@@ -24,6 +24,7 @@ exports.withAnalytics = (charge, param) => {
       analyticsId: charge.gatewayAccount.analyticsId,
       type: charge.gatewayAccount.type,
       paymentProvider: charge.gatewayAccount.paymentProvider,
-      amount: charge.amount
+      amount: charge.amount,
+      testingVariant: 'original'
```
Comment (maxcbc) [inline app/utils/analytics.js]: is this supposed to be hard coded?

Comment (tlwr) [reply app/utils/analytics.js]: default value that is overwritten by the `_merge`

File: app/utils/ab_test.js (hunk 2 above)
Comment (jonheslop) [reply app/utils/ab_test.js]: Is it just extra safe or is it silly? How should I refactor?

Comment (tlwr) [reply app/utils/ab_test.js]: The behaviour here is

- First calculate whether normally we would see the default
- Then if the showAbTest query param is truthy, force the ab test
- Then if the hideAbTest query param is truthy, force the default

This is superfluous, I guess we want a single query parameter to force/unforce

```
const showDefault = req.query.abTest
  ? req.query.abTest != 'yes'
  : sessionValue <= threshold
```

**Comment by heathd** [comment]
This change does not significantly increase the likelihood of people hitting the cookie size limit. It adds one cookie value for the session overall, not per-charge. Currently I believe that you have to make of the order of 20 payments to exceed the cookie size limit (just did a quick test and 5 payments had a cookie size of 1499 bytes, 10 payments 2139 bytes). 

We should revisit the cookie size issue and see if we can get it to fail more gracefully rather than with a 500 error.

I don't think concerns about cookie size should prevent this being merged.

If others have reviewed the code I'm happy for this to be merged.

**Pull request merged**